internal/staging_lockfile/staging_lockfile.go

```diff
@@ -0,0 +1,173 @@
+)
+
+// AssertLocked checks if the lock is currently held and panics if it's not.
+func (l *StagingLockFile) AssertLocked() {
+	// DO NOT provide a variant that returns the value of l.locked.
+	//
+	// If the caller does not hold the lock, l.locked might nevertheless be true because another goroutine does hold it, and
+	// we can’t tell the difference.
+	//
+	// Hence, this “AssertLocked” method, which exists only for sanity checks.
+
+	// The caller is expected to hold the lock, and in that case l.locked is constant true
+	// with no possible writers.
+	// If the caller does not hold the lock, we are violating the locking/memory model anyway, and accessing the data
+	// without the lock is more efficient for callers, and potentially more visible to lock analysers for incorrect callers.
+	if !l.locked {
+		panic("internal error: lock is not held by the expected owner")
+	}
+}
+
+// tryAcquireLockForFile attempts to acquire a lock for the specified file path.
+// It first checks if the lock is already in use by another thread.
+// If the lock is not in use, it creates a new StagingLockFile, opens the file, and tries to lock it.
+// If successful, it adds the StagingLockFile to the global map of staging lock files.
+// If the lock is already in use, it returns an error.
+func tryAcquireLockForFile(cleanPath string) (*StagingLockFile, error) {
+	stagingLockFileLock.Lock()
+	defer stagingLockFileLock.Unlock()
+
+	if stagingLockFiles == nil {
+		stagingLockFiles = make(map[string]*StagingLockFile)
+	}
+
+	if _, ok := stagingLockFiles[cleanPath]; ok {
+		return nil, fmt.Errorf("lock is used already with other thread %q", cleanPath)
+	}
+
+	lockFile := &StagingLockFile{
+		file:   cleanPath,
+		locked: false,
+	}
+
+	fd, err := rawfilelock.OpenLock(lockFile.file, false)
+	if err != nil {
+		return nil, err
+	}
+	lockFile.fd = fd
+
+	if err = rawfilelock.TryLockFile(lockFile.fd, rawfilelock.WriteLock); err != nil {
+		rawfilelock.CloseHandle(lockFile.fd) // This is safe because we hold stagingLockFileLock so we are the only possible holder of the lock within this process.
+		return nil, fmt.Errorf("failed to acquire lock on %q: %w", cleanPath, err)
+	}
+
+	lockFile.locked = true
+
+	stagingLockFiles[cleanPath] = lockFile
+	return lockFile, nil
+}
+
+// UnlockAndDelete releases the lock, removes the associated file from the filesystem,
+// and removes this StagingLockFile from the global map of StagingLockFile.
+//
+// WARNING: After this operation, the StagingLockFile becomes invalid for further use as the file field is cleared.
+func (l *StagingLockFile) UnlockAndDelete() error {
+	stagingLockFileLock.Lock()
+	defer stagingLockFileLock.Unlock()
+
+	if !l.locked {
+		// Panic when unlocking an unlocked lock. That's a violation
+		// of the lock semantics and will reveal such.
+		panic("calling Unlock on unlocked lock")
+	}
+
+	delete(stagingLockFiles, l.file)
+	l.locked = false
+
+	defer func() {
+		rawfilelock.UnlockAndCloseHandle(l.fd)
```

Comment:
```suggestion
		// It’s important that this happens while we are still holding stagingLockFileLock, to ensure
		// that no other goroutine has l.file open = that this close is not unlocking the lock under any
		// other goroutine.
		rawfilelock.UnlockAndCloseHandle(l.fd)
```
per https://github.com/containers/storage/pull/2336#discussion_r2110139525 .